Rust - add validation for typed union accessors

Tommytrg opened the pull request:
Follow @rw suggestion to fix #5055 that was discussed in #4967.

rw commented on src/idl_gen_rust.cpp:
```diff
@@ -294,7 +294,7 @@ class RustGenerator : public BaseGenerator {
   bool generate() {
     code_.Clear();
-    code_ += "// " + std::string(FlatBuffersGeneratedWarning()) + "\n\n"; 
+    code_ += "// " + std::string(FlatBuffersGeneratedWarning()) + "\n\n";
```
@Tommytrg what is this line for?

Tommytrg replied:
Sorry, it was editor's autotrim plugin

rw commented:
@Tommytrg also please regenerate the code, if applicable, using `./generate_code.sh`

rw commented on src/idl_gen_rust.cpp:
```diff
@@ -1327,12 +1327,21 @@ class RustGenerator : public BaseGenerator {
-        code_ += "      None";
-        code_ += "    }";
+
+        if (field.required) {
```
This would return `None` even if the data is present -- not what we want!

rw commented on src/idl_gen_rust.cpp:
```diff
@@ -1327,12 +1327,21 @@ class RustGenerator : public BaseGenerator {
+          code_ += "      None";
+        } else {
+          code_ += "    if self.{{FIELD_NAME}}_type() == {{U_ELEMENT_ENUM_TYPE}} {";
+          code_ += "      self.{{FIELD_NAME}}().map(|u| "
```
This is where the conditional on `required` needs to be: if the field is `required`, don't do the `Option` `map`.

Tommytrg replied:
Now, If the field is required, it will skip the `.map` and return `Some(U_ELEMENT_TABLE_TYPE)`

Tommytrg commented:
> @Tommytrg also please regenerate the code, if applicable, using `./generate_code.sh`

Running `./generate_code.sh`:
```
../flatc: warning: GRPC interface generator not implemented for binary
../flatc: warning: GRPC interface generator not implemented for JavaScript
../flatc: warning: GRPC interface generator not implemented for Dart
../flatc: warning: GRPC interface generator not implemented for TypeScript
../flatc: warning: GRPC interface generator not implemented for C#
../flatc: warning: GRPC interface generator not implemented for Python
../flatc: warning: GRPC interface generator not implemented for Lobster
../flatc: warning: GRPC interface generator not implemented for Lua
../flatc: warning: GRPC interface generator not implemented for Rust
../flatc: warning: GRPC interface generator not implemented for PHP
../flatc: warning: GRPC interface generator not implemented for binary
../flatc: warning: GRPC interface generator not implemented for JavaScript
../flatc: warning: GRPC interface generator not implemented for Dart
../flatc: warning: GRPC interface generator not implemented for TypeScript
../flatc: warning: GRPC interface generator not implemented for C#
../flatc: warning: GRPC interface generator not implemented for Python
../flatc: warning: GRPC interface generator not implemented for Lobster
../flatc: warning: GRPC interface generator not implemented for Lua
../flatc: warning: GRPC interface generator not implemented for Rust
../flatc: warning: GRPC interface generator not implemented for PHP
```
Any ideas?

Tommytrg commented:
Now, If the field is `required`, it will skip the `.map` and return `Some(U_ELEMENT_TABLE_TYPE)`

rw commented:
@Tommytrg C++ cod looks good. You'll still need to regenerate the rust code. Your execution of `generate_code.sh` looks fine (the warnings are expected, because the different languages have different features)

rw commented:
@Tommytrg I discussed this with @aardappel and we've decided that this would benefit from updated the main test schema with a `required` `union` field so that we can test this situation. I can guide you in updating that schema and adding tests, or I can do that myself. Do you have a preference? It would make this PR more robust!

The line in the `monster_test.fbs` file I want to add is:
`  any_required:Any (id:48);`

rw commented:
@Tommytrg yes! You can create a new test `mod` for this, maybe call it: `required_fields`.

Tommytrg commented:
I've added a `required_fields` mod and I've fixed almost all the tests that have broken with the new `any_required` field. I couldn't fix the tests which use `monsterdata_test.mon` file. How is this file generated?

Tommytrg commented:
@rw?

rw commented:
@Tommytrg Sorry I took a long time to get back to you -- this PR looks like it's on track! One change that would make merging this a lot easier would be to put the new schema field into a separate flatbuffers testing schema, not the main `monster.fbs`. I spoke with @aardappel and he pointed out that putting a new required field in the main testing schema will cause all the other language ports to break.

We do this already with other language features, such as union vectors. The steps will be:

1) Create a new test schema in the `tests/` directory, perhaps called "required_union.fbs". Here's an example: https://github.com/google/flatbuffers/blob/master/tests/monster_extra.fbs

2) Add a line to `generate_code.sh` to generate only Rust code with this new schema. Here's an example: https://github.com/google/flatbuffers/blob/master/tests/generate_code.sh#L20

3) Update the test suite to import these new generated files.

Thanks again for your help and your patience, we appreciate your work.

Tommytrg commented:
Cool, I'll update the PR with that.

aardappel commented:
Well, presumably it breaks because you can't be adding required fields without breaking backwards compatibility..

I am not a fan of adding lots of small schemas for features either.. that's going to generate a ton of extra test files for all languages.

WiSaGaN commented:
Are we on track of including this in 1.11.0 release? That would help a lot with users with Rust code base a lot.

aardappel commented:
@Tommytrg can you move the change to the schema to a rust-only schema under `rust_usage_test` ? That will stop all the other languages from breaking for the moment, and then we can merge this.

@WiSaGaN I am working on making 1.11 happen as we speak, and this can definitely be included. I expect it to be ready in "a few weeks" depending on how things go.

Tommytrg commented:
@aardappel Of course, I would update the PR in the next days

krojew commented:
Any chance of this being merged? Just got hit by an error.

aardappel commented:
@Tommytrg can you make the change I requested above?
